internal/fs/type_cache_test.go

```diff
@@ -0,0 +1,421 @@
+		[]byte(contents))
+
+	ExpectEq(nil, err)
+	ExpectNe(nil, fileObject)
+
+	// Stat-call with file object. It should
+	// pass stat call, bypassing type-cache, as a file.
+	fi, err = os.Stat(path.Join(mntDir, name1))
+
+	ExpectEq(nil, err)
+	AssertNe(nil, fi)
+	ExpectFalse(fi.IsDir())
+
+	// Create a directory object in GCS with same name as the file object.
+	dirObject, err := storageutil.CreateObject(
+		ctx,
+		bucket,
+		name1+"/",
+		[]byte(contents))
+
+	ExpectEq(nil, err)
+	ExpectNe(nil, dirObject)
+
+	// Stat-call with directory object. It should
+	// pass stat call, bypassing type-cache, as a directory.
+	// It works because no entries are inserted in type-cache
+	// in this case.
+	fi, err = os.Stat(path.Join(mntDir, name1) + "/")
+
+	ExpectEq(nil, err)
+	AssertNe(nil, fi)
+	ExpectTrue(fi.IsDir())
+}
+
+// //////////////////////////////////////////////////////////////////////
+// Tests for TypeCacheTestWithMaxEntries1
+// //////////////////////////////////////////////////////////////////////
+func (t *TypeCacheTestWithMaxSize1MB) TestSizeBasedEviction() {
+	const name1 = "foo"
+	const contents = "taco"
+	contentInBytes := []byte(contents)
+	var fi fs.FileInfo
+	var err error
+	objectNameTemplate := "object%06d" // 12
+	for i := 0; i < 93; i++ {
+		objectNameTemplate += "abcdefjhij" // +930
+	}
+	nameOfIthObject := func(i int) string {
+		return fmt.Sprintf(objectNameTemplate, i)
+	}
+	objectNameSample := nameOfIthObject(0)
+	perEntrySize := 80 + len(objectNameSample)
+
+	// Initially, without any existing object, type-cache
+	// should not contain any entry and os.Stat should fail.
+	_, err = os.Stat(path.Join(mntDir, name1))
+
+	ExpectNe(nil, err)
+
+	// Create a file object in GCS.
+	fileObject, err := storageutil.CreateObject(
+		ctx,
+		bucket,
+		name1,
+		[]byte(contents))
+
+	ExpectEq(nil, err)
+	ExpectNe(nil, fileObject)
+
+	// Stat-call with first file object. It should
+	// pass stat call, through type-cache as a file.
+	fi, err = os.Stat(path.Join(mntDir, name1))
+
+	ExpectEq(nil, err)
+	AssertNe(nil, fi)
+	ExpectFalse(fi.IsDir())
+
+	// Create a directory object in GCS with same name as the first file object.
+	dirObject, err := storageutil.CreateObject(
+		ctx,
+		bucket,
+		name1+"/",
+		contentInBytes)
+
+	ExpectEq(nil, err)
+	ExpectNe(nil, dirObject)
+
+	// Stat-call with the directory object. It should
+	// fail as there is currently an entry for the first
+	// file object, which has the same name.
+	_, err = os.Stat(path.Join(mntDir, name1) + "/")
+
+	ExpectNe(nil, err)
+	ExpectThat(err, oglematchers.Error(oglematchers.HasSubstr("not a directory")))
+
+	// type-cache-max-size-mb = 1MiB.
+	// let's add another 1MiB/perEntrySize entries to evict
+	// the first file object ("foo") entry.
+	numObjectsToBeInserted := int(util.MiBsToBytes(1)) / perEntrySize
+
+	for i := 0; i < numObjectsToBeInserted; i++ {
+		name := nameOfIthObject(i)
+
+		// Create a new file object in GCS.
+		fileObject, err = storageutil.CreateObject(
```

Comment:
As discussed offline, if we can parallelize the creation.

Reply:
done